### PR TITLE
[NeMo-UX] Adding MegatronParallel

### PR DESCRIPTION
**Pull request by cuichenx**
NeMo-UX PR 1, add megatron_parallel.py and associated tests

### CODE REVIEW

File: nemo/lightning/megatron_parallel.py [hunk 1]
```diff
@@ -0,0 +1,975 @@
...
+        precision_plugin: Optional[PrecisionPluginProtocol] = None,
+        callbacks: Optional["CallbackConnector"] = None,
+        data_step: Optional[Callable[[Iterator[DataT]], DataT]] = None,
+        forward_step: Optional[Callable[[nn.Module, DataT], Tensor]] = None,
+        loss_reduction: Optional[Callable[[nn.Module], "MegatronLossReduction"]] = None,
+        vp_size: Optional[int] = None,
+        cpu: bool = False,
+    ) -> None:
+        from apex.transformer.tensor_parallel.layers import set_defaults_if_not_set_tensor_model_parallel_attributes
+        from megatron.core import mpu
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:113]: Can you use parallel_state instead of mpu?

File: nemo/lightning/megatron_parallel.py [hunk 2]
```diff
@@ -0,0 +1,975 @@
...
+        pipeline: Union[nn.Module, Iterable[nn.Module]],
+        precision_plugin: Optional[PrecisionPluginProtocol] = None,
+        callbacks: Optional["CallbackConnector"] = None,
+        data_step: Optional[Callable[[Iterator[DataT]], DataT]] = None,
+        forward_step: Optional[Callable[[nn.Module, DataT], Tensor]] = None,
+        loss_reduction: Optional[Callable[[nn.Module], "MegatronLossReduction"]] = None,
+        vp_size: Optional[int] = None,
+        cpu: bool = False,
+    ) -> None:
+        from apex.transformer.tensor_parallel.layers import set_defaults_if_not_set_tensor_model_parallel_attributes
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:112]: We want to remove dependency on apex, can this temporarily go in nemo? Ultimately in mcore?

File: nemo/lightning/megatron_parallel.py [hunk 3]
```diff
@@ -0,0 +1,975 @@
...
+            forward_step_func = self.wrapped_forward_step(
+                _forward_step, data_step=_data_step, loss_reduction=loss_reduction, context=context,
+            )
+        else:
+            forward_step_func = _forward_step
+
+        self.callbacks.event("on_megatron_step_start", **context)
+        self.callbacks.event("on_megatron_microbatches_start", **context)
+
+        # TODO @akhattar: add num_micro_batches_with_partial_activation_checkpoints when ready
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:223]: do we need this comment still?

File: nemo/lightning/megatron_parallel.py [hunk 4]
```diff
@@ -0,0 +1,975 @@
...
+            num_microbatches=_num_microbatches,
+        )
+
+        context["microbatch_outputs"] = microbatch_outputs
+
+        self.callbacks.event("on_megatron_microbatches_end", **context)
+
+        if microbatch_outputs:
+            self.callbacks.event("on_megatron_reduce_microbatches_start", **context)
+            # Taken from: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L534
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:240]: Do we need this comment?

File: nemo/lightning/megatron_parallel.py [hunk 5]
```diff
@@ -0,0 +1,975 @@
...
+
+            # TODO: Can this lead to issues?
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:242]: Did we figure this out?

File: nemo/lightning/megatron_parallel.py [hunk 6]
```diff
@@ -0,0 +1,975 @@
...
+            context["loss_mean"] = loss_mean
+            self.callbacks.event("on_megatron_reduce_microbatches_end", **context)
+        else:
+            # we're not on the last pipeline stage so no losses
+            if forward_only:
+                loss_mean = cast(torch.Tensor, [])
+            else:
+                loss_mean = torch.tensor(0.0).cuda()
+
+        # TODO: How to handle all-reduce callbacks?
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:256]: do we need this comment?

File: nemo/lightning/megatron_parallel.py [hunk 7]
```diff
@@ -0,0 +1,975 @@
...
+    we need to interact with the data iterator multiple times
+    for each microbatch step. Instead of incorporating this
+    logic into the data loader, we cache the iterator's output
+    to the first model chunk and reuse it in the other model
+    chunks.
+    """
+    if not isinstance(model, list) or len(model) == 1:
+        return data_iterator  # TODO @tmoon: Remove
+        # TODO @tmoon: Use once available in Megatron-LM
+        # return DataIteratorList([data_iterator])
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:838]: Can we clean this up?

File: nemo/lightning/megatron_parallel.py [hunk 8]
```diff
@@ -0,0 +1,975 @@
...
+                proxy.cache.put(val)
+            return val
+
+    # Make list of iterator wrappers
+    iters = [CachingIterator(data_iterator)]
+    while len(iters) < len(model):
+        iters.append(iters[0].make_proxy())
+    return iters  # TODO @tmoon: Remove
+    # TODO @tmoon: Use once available in Megatron-LM
+    # return DataIteratorList(iters)
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:881]: can we clean this up?

File: nemo/lightning/megatron_parallel.py [hunk 9]
```diff
@@ -0,0 +1,975 @@
...
+        self.validation_step = validation_step
+        self.val_drop_last = val_drop_last
+
+    def forward(
+        self, batch: Dict[str, torch.Tensor], forward_out: torch.Tensor
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        """Taken from:
+        https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L951-L976 .
+        """
+        from megatron.core import mpu, parallel_state
```
Comment (ericharper) [inline nemo/lightning/megatron_parallel.py:896]: can just use parallel_state?

File: nemo/lightning/megatron_parallel.py (hunk 2 above)
Comment (marcromeyn) [reply nemo/lightning/megatron_parallel.py:112]: Yeah makes sense. Let's do that in a separate PR so we can update it everywhere in NeMo.

File: nemo/lightning/megatron_parallel.py (hunk 8 above)
Comment (marcromeyn) [reply nemo/lightning/megatron_parallel.py:879]: This is copied from elsewhere in NeMo. If we clean this up, let's do this in a separate PR so we can do it everywhere.

File: nemo/lightning/megatron_parallel.py (hunk 7 above)
Comment (marcromeyn) [reply nemo/lightning/megatron_parallel.py:831]: Let's clean this up everywhere soon, since this is copied from elsewhere in NeMo.

**Review by marcromeyn — APPROVED** [review]

**Pull request merged**